### DISCUSSION TITLE
Support partial cloning for large repos

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#21687] [Suggest an Idea]
**What would you like Renovate to be able to do?**

Support only partial cloning to save time and disk space on large repos where the majority of files do not need renovation.

**Did you already have any implementation ideas?**

There's at least two parts to how this needs to be done:

1. Which `git clone` options are most appropriate to use? GitLab has a good write-up on some [here](https://about.gitlab.com/blog/2020/03/13/partial-clone-for-massive-repositories/) and [here](https://docs.gitlab.com/ee/topics/git/partial_clone.html).

2. Where/how can Renovate be configured to achieve this? If it's an admin option only then it's easier to implement - because we know what commands to send before the clone. If it were to be in a repo config file, e.g. `renovate.json`, then it's a bit chicken and egg because today we naturally clone first in order to read that file. We could explore things like fetching the config file using API before cloning.

**Comment by rarkins** [comment]
Filtering out blobs seems like an easy one however it would be incompatible with some vendoring capabilities where binary archives need committing. Maybe we could clone without blobs by default then do a full (shallow) clone if we find the config includes vendored files?

**Comment by cpsauer** [comment]
Circling back to add thoughts from our use case ([issue mentioned above](https://github.com/renovatebot/app-support/issues/102))

On our side, we're delighted if (1) things work and (2) we're as little of an additional burden as possible to you, @rarkins. It's not important to me that it runs faster than that ~8 minutes, unless that's making things too costly to run for you. That said, if it does feel worth it, we'd probably be a model use case for partial cloning: We're only looking to update 1 bazel WORKSPACE file, and the repo's ~9GB shallow, since we've got a heavy combo of 3D assets and ML.

A divergent thought, maybe outside the box of minimizing download:
To avoid the failure case we hit, what about using the Github API to get the size of the repo for the purposes of sizing the disk? Or (worse, but maybe easier) auto enlarging--up to some limit--on disk space failure?

Or, if the repo's big, we could provide the paths to be scanned up front, either prompted through the signup UI or through a first PR. And then the config and indicated files could be selectively downloaded using sparse [checkout/selective cloning (GitHub article I liked, in addition to yours)](https://github.blog/2020-01-17-bring-your-monorepo-down-to-size-with-sparse-checkout/).

Those are my 2 cents!

**Comment by rarkins** [comment]
It would be good if this feature was controlled by our existing `includePaths` config option, although we may need to change the syntax to more easily map to git selective cloning. Then we'd need to eager fetch the `renovate.json` similar to how we do with `optimizeForDisabled` today.

**Comment by rarkins** [comment]
Some more research: https://github.blog/2020-12-22-git-clone-a-data-driven-study-on-cloning-behaviors/

**Comment by rarkins** [comment]
The chicken and egg problem can be solved by our more recent use of platform APIs to fetch single files. That way we can detect and fetch the renovate config file prior to cloning, in turn allowing for options to be configured which speed up cloning through sparse checkout and/or partial cloning (best described in the GitHub blog post last year).

One challenge is: the "cost" of expensive clones may be invisible most repo users, so they may be not only uninterested in optimising their config but also unaware. There are naturally *some* direct benefits to them in Renovate being more responsive though.

**Comment by rarkins** [comment]
This was the other article I was reading again: https://github.blog/2020-01-17-bring-your-monorepo-down-to-size-with-sparse-checkout/

We could maybe just add a new config option like gotSparseCheckoutSet and pass it through

**Comment by rarkins** [comment]
More: https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/

**Comment by rarkins** [comment]
> Blobless partial clones are particularly effective if you are using Git’s sparse-checkout feature to reduce the size of your working directory. The combination greatly reduces the number of blobs you need to do your work. Sparse-checkout does not reduce the required data transfer for shallow clones.

Treeless also performs badly with origin resets:

> The two partial clone options have drastically different behavior in the git fetch and git reset --hard origin/$branch sequence. Fetching from blobless partial clones increases the reset command by a small, measurable way, but not enough to make a huge difference from the user perspective. In contrast, fetching from a treeless partial clone causes significant more time because the server needs to send all trees and blobs reachable from a commit’s root tree in order to satisfy the git reset --hard origin/$branch command.

(I'm not 100% sure we need origin resets though)

**Comment by rarkins** [comment]
Combined example:

```

$ git clone --filter=blob:none --no-checkout https://github.com/derrickstolee/sparse-checkout-example
Cloning into 'sparse-checkout-example'...
Receiving objects: 100% (373/373), 75.98 KiB | 2.71 MiB/s, done.
Resolving deltas: 100% (23/23), done.
 
$ cd sparse-checkout-example/
 
$ git checkout main
remote: Enumerating objects: 2, done.
remote: Counting objects: 100% (2/2), done.
remote: Compressing objects: 100% (2/2), done.
remote: Total 3 (delta 0), reused 1 (delta 0), pack-reused 1
Receiving objects: 100% (3/3), 1.41 KiB | 1.41 MiB/s, done.
Already on 'main'
Your branch is up to date with 'origin/main'.

$ git sparse-checkout set client/android
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 26 (delta 0), reused 1 (delta 0), pack-reused 23
Receiving objects: 100% (26/26), 985.91 KiB | 13.69 MiB/s, done.
```

**Comment by rarkins** [comment]
https://github.blog/2021-11-10-make-your-monorepo-feel-small-with-gits-sparse-index/

**Comment by rarkins** [comment]
From the [2.35 release notes](https://github.blog/2022-01-24-highlights-from-git-2-35/):

![image](https://user-images.githubusercontent.com/6311784/150940486-f637da40-5d3f-41dd-9a69-de09c13ddeba.png)
